feat/new-lessons-on-js-promises (#36045)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/handle-a-fulfilled-promise-with-then.english.md

---
id: 5cdafbd72913098997531681
title: Handle a Fulfilled Promise with then
challengeType: 1
---

## Description
<section id='description'>

> Promises are most useful when you have a process that takes an unknown amount of time in your code, something asynchronous. Often, a server request. When you make a server request it takes some amount of time, and after it completes you usually want to do something with the response from the server. This can be achieved by using the <code>then</code> method. The <code>then</code> method is executed immediately after your promise is fulfilled with <code>resolve</code>. Here’s an example:

Reviewer: The commas before "something asynchronous" was not correct.  I have provided an alternative way of handling this.
```suggestion
Promises are most useful when you have a process that takes an unknown amount of time in your code (i.e. something asynchronous). Often, a server request. When you make a server request it takes some amount of time, and after it completes you usually want to do something with the response from the server. This can be achieved by using the <code>then</code> method. The <code>then</code> method is executed immediately after your promise is fulfilled with <code>resolve</code>. Here’s an example:
```

Author: Are you sure those commas aren't right? Either way - your method works fine for me.

```js
myPromise.then(result => {
  // do something with the result.
});
```

<code>result</code> comes from the argument given to the <code>resolve</code> method.
</section>

## Instructions
<section id='instructions'>
Use the <code>then</code> method on your promise. Log its argument to the console.
</section>

## Tests
<section id='tests'>

```yml
tests:
  - text: You should call the <code>then</code> method on the promise.

>     testString: assert(code.match(/(makeServerRequest|\))\.then\(\s*(\s*function\s*\(\s*result\s*\)\s*|\s*result\s*=>\s*|\s*\(\s*result\s*\)\s*=>\s*){[^}]*}\s*\)/g));

Reviewer: I don't have an immediate code suggestion here, but if the text of the test is going to be "You should call the `then` method on the promise", then it should really only be checking if at a minimum `.then()` is used.  The current test focuses on the construction of the callback function.  Maybe another test should be added to test for the existence of using some kind of callback function.

Author: I can try and come up with something a little more specific here.

Reviewer: LGTM

  - text: You should log the <code>then</code> argument (<code>result</code>) to the console.

>     testString: assert(code.match(/{[^c]*console\.log\(\s*result\s*\)[^}]*}/g));

Reviewer: Related to my comment above, Since what the user decides to choose for the parameter name of the callback function (it does not have to be "result"), then why not figure out a way to actually have the test make a call and validate something is logged to the console.  @ValeraS Any suggestions how to do something like this with the tests or another creative way to validate the then method is used to do something specific?

Also, the following would be a completely valid way of using console.log of the result but the regex is not flexible enough to handle it.
```
makeServerRequest.then(result => console.log(result));
```

Author: Yes, that would be ideal - I thought about that, but didn't figure anything out and decided to just regex it for now. I think if the regex can accommodate the various syntax it should be fine. We could force the campers to use 'result' as the parameter name or maybe not check the parameter name in the regex. I'll take another look here and see if I can figure something out.

Reviewer: You could use `(\w)` instead of `result` in the regex and then obtain the name of the parameter used using `match` in order to reference in a later test.

---
id: 5cdafbd72913098997531681
title: Handle a Fulfilled Promise with then
challengeType: 1
---

## Description
<section id='description'>
Promises are most useful when you have a process that takes an unknown amount of time in your code (i.e. something asynchronous), often a server request. When you make a server request it takes some amount of time, and after it completes you usually want to do something with the response from the server. This can be achieved by using the <code>then</code> method. The <code>then</code> method is executed immediately after your promise is fulfilled with <code>resolve</code>. Here’s an example:

```js
myPromise.then(result => {
  // do something with the result.
});
```

<code>result</code> comes from the argument given to the <code>resolve</code> method.
</section>

## Instructions
<section id='instructions'>

> Add the <code>then</code> method to your promise. Use <code>result</code> as the argument of its callback function and log <code>result</code> to the console.

Reviewer 2: ```suggestion
Add the <code>then</code> method to your promise. Use <code>result</code> as the parameter of its callback function and log <code>result</code> to the console.
```

Reviewer: @moT01 I thought you already fixed this in a previous commit.  Maybe when you rebased and fixed all the merge conflicts, it cropped back up.

---
id: 5cdafbd72913098997531681
title: Handle a Fulfilled Promise with then
challengeType: 1
---

## Description
<section id='description'>
Promises are most useful when you have a process that takes an unknown amount of time in your code, something asynchronous. Often, a server request. When you make a server request it takes some amount of time, and after it completes you usually want to do something with the response from the server. Introducing the <code>then</code> method. The <code>then</code> method is executed immediately after your promise is fulfilled with <code>resolve</code>. Here’s an example:

```js
myPromise.then(result => {
  // do something with the result.
});
```

<code>result</code> comes from the argument given to the <code>resolve</code> method.
</section>

## Instructions
<section id='instructions'>
Use the <code>then</code> method on your promise. Log its argument to the console.
</section>

## Tests
<section id='tests'>

```yml
tests:

>   - text: You should properly use the <code>then</code> method.

Reviewer 3: ```suggestion
  - text: You should call the <code>then</code> method on the promise.
```
This gives them a little more of a hint.

    testString: assert(code.match(/(makeServerRequest|\))\.then\(\s*(\s*function\s*\(\s*result\s*\)\s*|\s*result\s*=>\s*|\s*\(\s*result\s*\)\s*=>\s*){[^}]*}\s*\)/g));
  - text: You should log the <code>then</code> argument (<code>result</code>) to the console.
    testString: assert(code.match(/{[^c]*console\.log\(\s*result\s*\)[^}]*}/g));
```

</section>

## Challenge Seed
<section id='challengeSeed'>
<div id='js-seed'>

```js
const makeServerRequest = new Promise((resolve, reject) => {
  const responseFromServer = true;
	
  if(responseFromServer) {
    resolve("We got the data");
  } else {	
    reject("Data not received");
  }
});
```

</div>

> ## Challenge Seed

Reviewer 3: There shouldn't be a second seed.

<section id='challengeSeed'>
<div id='js-seed'>

```js
const makeServerRequest = new Promise((resolve, reject) => {

>   const responseFromServer = true;

Reviewer 3: ```suggestion
  let responseFromServer;
  \\ the server responds and responseFromServer is set to true or false
```
Mirroring the change to the other challenge.

Author: For the .then and .catch lessons - if it isn't given a true or false value, resolve or reject won't be executed. Therefore, .then and .catch won't be executed. So I'm thinking something like this... 
```suggestion
\\ responseFromServer is set to true to represent a successful response from a server
let responseFromServer = true;
```

Reviewer 3: Yeah, that works.  I didn't like `const thing = true` followed by `if(thing)`, but the way you have it now gets across what's being abstracted away.

	
  if(responseFromServer) {
    resolve("We got the data");
  } else {	
    reject("Data not received");
  }
});
```

</div>
</section>

## Solution
<section id='solution'>

```js
const makeServerRequest = new Promise((resolve, reject) => {

>   const responseFromServer = true;

Reviewer 3: ```suggestion
  let responseFromServer;
  \\ the server responds and responseFromServer is set to true or false
```
Mirroring the change to the other challenge

Author: I will resolve this to remove some clutter. Whatever we decide on for the seed will be added here.
